### PR TITLE
feat(gossipsub): apply `max_transmit_size` to the published message

### PR DESCRIPTION
**Pull request by kalabukdima**
## Description

When trying to publish a message using gossipsub's `publish` method,
it should be possible to predict whether it will fit in the limit defined by
the `max_transmit_size` config option.
If this limit applies to the final protobuf payload, it's not possible to know
that in advance because the size of the added fields is not fixed.

This change makes the limit apply to the passed message size instead of the final wire size.

## Notes & open questions

This is a minor version change because it changes the meaning of the existing config option.
However, for the existing clients the limit will only become more permissive, so it shouldn't break anything.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/gossipsub/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 0.48.0
+
+- Apply `max_transmit_size` to the inner message instead of the final payload.
+  See [Discussion 5598](https://github.com/libp2p/rust-libp2p/discussions/5598).
```
Comment (dariusc93) [inline protocols/gossipsub/CHANGELOG.md]:
```suggestion
  See [PR 5642](https://github.com/libp2p/rust-libp2p/pull/5642).
```

**Review by dariusc93 — COMMENTED** [review]
LGTM, but we could probably have this be a patch bump instead of a minor since this could be treated more as a fix. Thoughts?

File: protocols/gossipsub/src/behaviour.rs [hunk 2]
```diff
@@ -583,6 +583,11 @@ where
             .data_transform
             .outbound_transform(&topic, data.clone())?;
 
+        // check that the size doesn't exceed the max transmission size
```
Comment (jxs) [inline protocols/gossipsub/src/behaviour.rs]:
```suggestion
        // check that the size doesn't exceed the max transmission size.
```

**Review by jxs — APPROVED** [review]
LGTM! thanks Dzmitry! @dariusc93 no worries, we also have https://github.com/libp2p/rust-libp2p/pull/5595 on the pipeline, so it's allright the CHANGELOG like this

**Review by dariusc93 — APPROVED** [review]

**Pull request merged**